src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -822,14 +822,54 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     res
   }
 
-  def typedAppliedTypeTree(tree: untpd.AppliedTypeTree)(implicit ctx: Context): AppliedTypeTree = track("typedAppliedTypeTree") {
+  def typedAppliedTypeTree(tree: untpd.AppliedTypeTree)(implicit ctx: Context): Tree = track("typedAppliedTypeTree") {
     val tpt1 = typed(tree.tpt)
-    val argPts =
-      if (ctx.mode is Mode.Pattern) tpt1.tpe.typeParams.map(_.info)
-      else tree.args.map(_ => WildcardType)
-    val args1 = tree.args.zipWithConserve(argPts)(typed(_, _)).asInstanceOf[List[Tree]]
-    // check that arguments conform to bounds is done in phase FirstTransform
-    assignType(cpy.AppliedTypeTree(tree)(tpt1, args1), tpt1, args1)
+    val tparams = tpt1.tpe.typeParams
+    var args = tree.args
+    if (tparams.isEmpty) {
+      ctx.error(d"${tpt1.tpe} does not take type parameters", tree.pos)
+      tpt1
+    }
+    else {
+      if (args.length != tparams.length) {
+        ctx.error(d"wrong number of type arguments for ${tpt1.tpe}, should be ${tparams.length}", tree.pos)
+        args = args.take(tparams.length)
+      }
+      def typedArg(arg: untpd.Tree, tparam: Symbol) = {
+        val arg1 = typed(arg, if (ctx.mode is Mode.Pattern) tparam.info else WildcardType)
+        adaptTypeArg(arg1, if (tparam.isCompleted) tparam.info else WildcardType)
+      }
+      val args1 = args.zipWithConserve(tparams)(typedArg(_, _)).asInstanceOf[List[Tree]]
+      // check that arguments conform to bounds is done in phase PostTyper
+      val tree1 = assignType(cpy.AppliedTypeTree(tree)(tpt1, args1), tpt1, args1)
+      if (tree1.tpe.isHKApply)
+        for (arg @ TypeBoundsTree(_, _) <- args1)
+          ctx.error("illegal wildcard type argument; does not correspond to type parameter of a class", arg.pos)
+          // The reason for outlawing such arguments is illustrated by the following example.
+          // Say we have
+          //
+          //    type RMap[A, B] = Map[B, A]
+          //
+          // Then
+          //
+          //    Rmap[_, Int]
+          //
+          // translates to
+          //
+          //    Lambda$I { type hk$0; type hk$1 = Int; type $apply = Map[$hk1, $hk0] } # $apply
+          //
+          // Let's call the last type T. You would expect that
+          //
+          //    Map[Int, String] <: RMap[_, Int]
+          //
+          // But that's not the case given the standard subtyping rules. In fact, the rhs reduces to
+          //
+          //    Map[Int, T # $hk0]
```

Comment:
But wildcard arguments to higher-kinded types are still disallowed,  I think it would be possible to allow them with a better implementation of `compareHK`, but I'm not sure.